File: Fortran/IBM/execute/F2018/cintrop_ts29113/optional/unit_tests/func/OptionalArg10f.f
!**********************************************************************
!*  ===================================================================
!*
!*  DATE                       : May 23, 2012
!*  ORIGIN                     : AIX Compiler Development,
!*
!*  PRIMARY FUNCTIONS TESTED   : C-interop OPTIONAL argument
!*
!*  SECONDARY FUNCTIONS TESTED : None
!*
!*  REQUIRED COMPILER OPTIONS  :
!*
!*  DESCRIPTION                : Calling a BIND(C) procedure from Fortran
!*                               where the procedure is defined in C and
!*                               contiguity checking for copy-in/out is
!*                               required.
!*                               - -qxlf2008=checkpresence is enabled and
!*                                 The actual arg is a pointer/allocatable
!*
!234567890123456789012345678901234567890123456789012345678901234567890

@process xlf2008(checkpresence)

      use, intrinsic :: iso_c_binding
      implicit none

      interface
      subroutine sub(arg, b) bind(c)
        import c_int,  c_float
        integer(c_int), optional :: arg(10)
        real(c_float),    optional :: b(10)
      end
      end interface

      integer, pointer :: arr(:), i
      real, pointer :: bb(:)
      allocate(arr(10), bb(10))
      arr = [(i,i=1,10,1)]
      bb = [(1.1*i, i=1, 10, 1)]

      call sub(arr)
      call sub()
      call sub(arr, bb)

      deallocate(bb)
      call sub(arr, bb)

      allocate(bb(10))
      bb = [(1.1*i, i=1, 10, 1)]
      deallocate(arr)
      call sub(arr, bb)

      deallocate(bb)
      call sub(arr, bb)

      end

File: Fortran/IBM/execute/F2018/cintrop_ts29113/optional/unit_tests/func/OptionalArg10f.c
/**********************************************************************
*  ===================================================================
*  XL Fortran Test Case                          IBM INTERNAL USE ONLY
*  ===================================================================
*
*  TEST CASE TITLE            : OptionalArg10f
*
*  PROGRAMMER                 : Bardia Mahjour
*  DATE                       : May 23, 2012
*  ORIGIN                     : AIX Compiler Development,
*                             : IBM Software Solutions Toronto Lab
*
*  PRIMARY FUNCTIONS TESTED   : C-interop OPTIONAL argument
*                                                   
*  SECONDARY FUNCTIONS TESTED : None 
*
*  DRIVER STANZA              : xlc
*  REQUIRED COMPILER OPTIONS  : 
*
*  DESCRIPTION                : Calling a BIND(C) procedure from Fortran
*                               where the procedure is defined in C and 
*                               contiguity checking for copy-in/out is
*                               required.
*                               - -qxlf2008=checkpresence is enabled and
*                                 The actual arg is a pointer/allocatable
*
************************************************************************/

#include <stdio.h>

void sub(int * a, float * b)
{
  int i;
  printf("--------------------\n");
  if (a == NULL)
  {
    printf("A: Not present!\n");
  }
  else
  {
    printf("A: ");
    for(i = 0; i < 10; i++)
      printf("%d ", a[i]);
    printf("\n");
  }
  if (b == NULL)
  {
    printf("B: Not present!\n");
  }
  else
  {
    printf("B: ");
    for(i = 0; i < 10; i++)
      printf("%.1f ", b[i]);
    printf("\n");
  }
  printf("--------------------\n");
}
